### PR TITLE
Document `ResourceProvider.Create` and `ResourceProvider.Read`

### PR DESCRIPTION
**Pull request by lunaris**
This commit fleshes out the documentation for the `Create` and `Read` methods of resource providers.

### CODE REVIEW

File: proto/pulumi/provider.proto [hunk 1]
```diff
@@ -593,35 +602,78 @@ message DiffResponse {
...
+    string id = 1;
+
+    // The URN of the resource being read.
+    string urn = 2;
+
+    // Any current state for the resource being read. This state should be sufficient to uniquely identify the resource.
+    google.protobuf.Struct properties = 3;
+
+    // Any current input properties for the resource being read. These will only be populated when the
+    // [](pulumirpc.ResourceProvider.Read) call is being made as part of a refresh operation.
```
Comment (flostadler) [inline proto/pulumi/provider.proto:653]: It would be great if we could document what indicates that this is a refresh

Comment (lunaris) [reply proto/pulumi/provider.proto:653]: Tracked in #17691

File: proto/pulumi/provider.proto [hunk 2]
```diff
@@ -133,11 +133,20 @@ service ResourceProvider {
...
     rpc Diff(DiffRequest) returns (DiffResponse) {}
-    // Create allocates a new instance of the provided resource and returns its unique ID afterwards.  (The input ID
-    // must be blank.)  If this call fails, the resource must not have been created (i.e., it is "transactional").
+
+    // `Create` provisions a new instance of the specified [(custom) resource](custom-resources). It returns a
+    // provider-assigned ID for the resource as well as the output properties that arose from the creation properties.
+    // Output properties are typically the union of the resource's input properties and any additional values that were
+    // computed or made available during creation.
+    //
+    // If creation fails, `Create` may return an [](pulumirpc.ErrorResourceInitFailed) error detail explaining why.
```
Comment (t0yv0) [inline proto/pulumi/provider.proto:142]: An idempotency discussion might be pertinent here, but maybe later as we have not figured it out. If a provider can signal that a resource is "already created" how can it do that.

There is definitely a partially created state resources enter, I think with pulumirpc.ErrorResourceInitFailed, they get tracked in state but it is still considered an error. I'm confused by that and would appreciate a page explaining that concept specifically, probably just to link here.

File: proto/pulumi/provider.proto [hunk 3]
```diff
@@ -133,11 +133,20 @@ service ResourceProvider {
...
+    // Moreover, if `Create` does return an error, it must be the case that the resource was *not* created (that is,
+    // `Create` can be thought of as transactional or atomic).
     rpc Create(CreateRequest) returns (CreateResponse) {}
-    // Read the current live state associated with a resource.  Enough state must be include in the inputs to uniquely
-    // identify the resource; this is typically just the resource ID, but may also include some properties.
+
+    // `Read` reads the current live state associated with a resource identified by the supplied state. The given state
```
Comment (t0yv0) [inline proto/pulumi/provider.proto:147]: This one will need a lot of work and is one of the most confusing bits of Pulumi for implementing providers.

More on it here:

https://github.com/pulumi/pulumi/issues/17650

We might have to go ahead and link all the relevant scenarios that invoke Read (two forms of import, refresh, and resource getters).

File: proto/pulumi/provider.proto [hunk 4]
```diff
@@ -594,35 +603,78 @@ message DiffResponse {
...
-    string name = 5; // the Pulumi name for this resource.
-    string type = 6; // the Pulumi type for this resource.
+    // The URN of the resource being created.
+    string urn = 1;
+
+    // The resource's input properties, to be set during creation. These should have been validated by a call to
+    // [](pulumirpc.ResourceProvider.Check).
+    google.protobuf.Struct properties = 2;
+
+    // A timeout in seconds that the caller is prepared to wait for the operation to complete.
```
Comment (t0yv0) [inline proto/pulumi/provider.proto:615]: Any preference on how the provider should indicate that the operation timed out?

File: proto/pulumi/provider.proto [hunk 5]
```diff
@@ -594,35 +603,78 @@ message DiffResponse {
...
+    double timeout = 3;
+
+    // True if and only if the request is being made as part of a preview/dry run, in which case the provider should not
```
Comment (t0yv0) [inline proto/pulumi/provider.proto:618]: I've always been curious what is the point of these preview calls, how are they used? Are they used to generate errors, or they influence the display? Or Both?

File: proto/pulumi/provider.proto (hunk 2 above)
Comment (lunaris) [reply proto/pulumi/provider.proto:142]: Tracked in #17691

File: proto/pulumi/provider.proto (hunk 3 above)
Comment (lunaris) [reply proto/pulumi/provider.proto:147]: Tracked in #17691

File: proto/pulumi/provider.proto (hunk 1 above)
Comment (t0yv0) [reply proto/pulumi/provider.proto:654]: > These will only be populated when 

Are we 100% sure? Resource getters and two forms of import do not populate this? Just making sure to check!

Also wonder if project Pangea ever sets these.

File: proto/pulumi/provider.proto (hunk 4 above)
Comment (lunaris) [reply proto/pulumi/provider.proto:615]: I need to look into what we actually do here. Will track in #17691

File: proto/pulumi/provider.proto [hunk 6]
```diff
@@ -594,35 +603,78 @@ message DiffResponse {
...
-    google.protobuf.Struct inputs = 4;     // the current inputs, if any (only populated during refresh).
-    string name = 5; // the Pulumi name for this resource.
-    string type = 6; // the Pulumi type for this resource.
+    // The ID of the resource to read.
+    string id = 1;
+
+    // The URN of the resource being read.
+    string urn = 2;
+
+    // Any current state for the resource being read. This state should be sufficient to uniquely identify the resource.
```
Comment (t0yv0) [inline proto/pulumi/provider.proto:650]: If more than one cloud resource matches, are there any guidance on what the provider should do, just err generically or a specific error? Or pick one?

File: proto/pulumi/provider.proto [hunk 7]
```diff
@@ -594,35 +603,78 @@ message DiffResponse {
...
+    // The name of the resource being read. This must match the name specified by the `urn` field, and is passed so that
+    // providers do not have to implement URN parsing in order to extract the name of the resource.
+    string name = 5;
+
+    // The type of the resource being read. This must match the type specified by the `urn` field, and is passed so that
+    // providers do not have to implement URN parsing in order to extract the type of the resource.
+    string type = 6;
 }
 
+// `ReadResponse` is the type of responses sent by a [](pulumirpc.ResourceProvider.Read) call. A `ReadResponse` contains
```
Comment (t0yv0) [inline proto/pulumi/provider.proto:666]: We must document how to indicate that a resource could not be found in the cloud. It is super common.

File: proto/pulumi/provider.proto (hunk 6 above)
Comment (lunaris) [reply proto/pulumi/provider.proto:650]: Tracked in #17691

File: proto/pulumi/provider.proto [hunk 8]
```diff
@@ -594,35 +603,78 @@ message DiffResponse {
...
-    string id = 1;                         // the ID of the resource read back (or empty if missing).
-    google.protobuf.Struct properties = 2; // the state of the resource read from the live environment.
-    google.protobuf.Struct inputs = 3;     // the inputs for this resource that would be returned from Check.
+    // The ID of the read resource.
+    string id = 1;
+
+    // The output properties of the resource read from the live environment.
+    google.protobuf.Struct properties = 2;
+
+    // Output-derived input properties for the resource. These are returned as they would be returned from a
```
Comment (t0yv0) [inline proto/pulumi/provider.proto:675]: Most provides are in violation of this expectation here, sadly.

File: proto/pulumi/provider.proto (hunk 7 above)
Comment (lunaris) [reply proto/pulumi/provider.proto:666]: Yep I've added a general "what are the error cases?" point to #17691. I have a lot of digging to do xD

**Comment by t0yv0** [comment]
This is great! Not to stop this, there's just more nitty-gritty things that come up in debugging providers! Left some notes.

**Review by julienp — APPROVED** [review]

**Pull request closed without merging**